src/Nonlinear/Nonlinear.jl

```diff
@@ -0,0 +1,515 @@
+module Nonlinear
```

Comment:
Yes, this is a little complicated. Most of the code in Nonlinear, you or I wrote.

ReverseDiffSparse was MIT: https://github.com/mlubin/ReverseDiffSparse.jl/blob/master/LICENSE.md
The `_Derivatives` submodule was moved to JuMP here: https://github.com/jump-dev/JuMP.jl/pull/1197, and so I guess implicitly under MPL? But we never copied the license or asked the authors.

Since then, the only people to commit to `_Derivatives` has been me and you, and https://github.com/jump-dev/JuMP.jl/commit/67aaf56ea545684cb73c92ee86c359f65e9157af, which I don't think requires permission.

There have been a few more commits to `nlp.jl`, the largest that made its way here was https://github.com/jump-dev/JuMP.jl/pull/2196. The others were at the JuMP level and remained in `nlp.jl`.

Reply:
That sounds fine. We do need permission from the authors of https://github.com/jump-dev/JuMP.jl/pull/2196 to relicence their code.

Reply:
I've gone through the blame, and I think this is the only case. Other potentials are:
 * https://github.com/jump-dev/JuMP.jl/pull/1551, which I've since refactored
 * https://github.com/jump-dev/JuMP.jl/pull/1416, which introduced `@elapsed` that has now been removed
Most of the code hasn't been touched for 4-8 years, and the other contributors to `nlp.jl` have either already had their contributions removed in the MPB -> MOI transition, or touch JuMP-facing parts that remained in JuMP.

Reply:
@amontoison, we're moving some code from JuMP to MOI, and this involves a relicensing the code you wrote in https://github.com/jump-dev/JuMP.jl/pull/2196 from MPL2.0 to MIT. Can you please reply in the affirmative if you are okay with this.

Reply:
Yes @odow, it's fine for me !

Reply:
Great! Your PR https://github.com/jump-dev/JuMP.jl/pull/2239 is on my list to tackle once we get this all sorted :smile:

Reply:
The `JF(x)' * v` operator should be easy to add too, we "just" need to derivate the dot product `<F(x), v>`.
The most difficult part is to understand the undocumented code of `nlp.jl`.

Reply:
This PR cleans up a lot of the undocumented code: https://github.com/jump-dev/MathOptInterface.jl/pull/1803